zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -202,21 +210,32 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
                 )
                 .to_word(),
             );
-        });
-        cb.require_equal(
-            "tx nonce equivalence",
-            tx.nonce.expr(),
-            create.caller_nonce(),
-        );
-
-        // 1. Handle contract creation transaction.
-        cb.condition(tx.is_create.expr(), |cb| {
             cb.keccak_table_lookup(
                 create.input_rlc(cb),
                 create.input_length(),
                 caller_nonce_hash_bytes.to_word(),
             );
 
+            let length = cb.tx_context(tx_id.expr(), TxContextFieldTag::CallDataLength, None);
+            let rlc_acc = cb.query_cell_phase2();
+            cb.copy_table_lookup(
+                WordLoHi::from_lo_unchecked(tx_id.expr()),
+                CopyDataType::TxCalldata.expr(),
+                WordLoHi::zero(),
+                CopyDataType::RlcAcc.expr(),
+                0.expr(),
+                length.expr(),
+                0.expr(),
+                length.expr(),
+                rlc_acc.expr(),
+                0.expr(),
```

Comment:
I checked the errors and noticed there were quite a few things missing!  I think it's not trivial to be aware of the missing things without being familiar with the entire system :)

Take a look at this commit where I added the missing things in order to make the tests pass https://github.com/privacy-scaling-explorations/zkevm-circuits/commit/2291146244cb98d486bf5321eb65b42aa38bb9b1

Here's a list of changes:
- The cells `calldata_length` and `calldata_rlc` in BeginTx where not being assigned, so I added them to the gadget and assigned them in `BeginTx::assign_exec_step`
- The keccak table didn't have the input for tx.call_data when the tx is creation, so I added it via the bus-mapping
- The copy table didn't have the input for copying the tx.call_data into rlc when the tx is creation, so I added it via the bus-mapping
- The copy table only contains an entry in case there's at least 1 byte copied.  If `tx.call_data.len() == 0` there will be no entry in the copy table, so the lookup will fail.  I added a conditional on the copy table lookup for the case `tx.call_data.len() > 0`, and otherwise I directly set `call_data_rlc = 0`

Reply:
May I ask - is it necessary add `calldata_length`, `calldata_rlc`, etc. to the gadget, and therefore assign them. Why wouldn't it work the way I made it before? Or the cells "were not there" before? 

For example what's the fundamental difference between this:
```
let length = cb.tx_context(tx_id, TxContextFieldTag::CallDataLength, None);
```
and having a separate `calldata_length` field/cell in the gadget, and doing the lookup:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/2291146244cb98d486bf5321eb65b42aa38bb9b1/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L226-L231

---

> The copy table only contains an entry in case there's at least 1 byte copied. If tx.call_data.len() == 0 there will be no entry in the copy table, so the lookup will fail. I added a conditional on the copy table lookup for the case tx.call_data.len() > 0, and otherwise I directly set call_data_rlc = 0

Yes, I was going through Scroll's begin_tx and they have similar thing; I forgot to add the same check here, sorry.

Thanks for help, I'll merge branches and push it here.

Reply:
> May I ask - is it necessary add `calldata_length`, `calldata_rlc`, etc. to the gadget, and therefore assign them. Why wouldn't it work the way I made it before? Or the cells "were not there" before?

If you don't add these cells to the gadget, they will not be part of the config, and if they are not part of the config, you don't have the reference to do the witness assignment, and then they will be assigned to 0, so the constraints will fail.  It was working before because these two cells didn't exists before.

> For example what's the fundamental difference between this:
> 
> ```
> let length = cb.tx_context(tx_id, TxContextFieldTag::CallDataLength, None);
> ```
> 
> and having a separate `calldata_length` field/cell in the gadget, and doing the lookup:
> 
> https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/2291146244cb98d486bf5321eb65b42aa38bb9b1/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L226-L231

The first option is always nicer, but it does two things:
1. Allocate the cell
2. Perform the lookup with that cell

In this case, we need to allocate the cell unconditionally because we want to store it in the config to later on use it for witness assignment, but the lookup must be conditional!

Another option would be doing something like this:
```rust
let length = cb.condition(tx.is_create.expr(), |cb| {
    // [...]
    let length = cb.tx_context(tx_id, TxContextFieldTag::CallDataLength, None);
    // [...]
    length
};
```
That would be equivalent.

Reply:
Thanks a lot for explanation